Default user registration

alintreznai opened the pull request:
This PR implements a default registration proceses for new users.

Commit:
Default user registration

Commit:
Structure cleanup

Commit:
Refactor default user registration

Commit:
Server side validation for registration forms

Commit:
Cleanup registration class usage

Commit:
Simplified code

sztamas reviewed (commented):
Patient registration page problems:

- header should be Patient registration (instead of User registration)
- the gender field has list item "bullets" on
- the gender field error message appears in the wrong place (on the first list item instead above of below the whole list)
- you can submit the form without setting the State
- the passwords are lost (have to retyped again) if you fail on server-side validation
- the recaptcha isn't tested server-side? If you wait until the recaptcha expires you can submit the form without the recaptcha test again, so I'm assuming it isn't tested at all server-side.

sztamas commented on rdrf/rdrf/forms/registration_forms.py:
```diff
@@ -0,0 +1,122 @@
+
+def _countries():
+    countries = sorted(pycountry.countries, key=attrgetter('name'))
+    result = [_tuple("-1", "Country")]
```
Shouldn't this be "" instead of "-1" so that the required validator kick in?

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/templates/registration/registration_base.html:
```diff
@@ -53,18 +53,29 @@
                 var state_lookup_url = "{% url 'v1:state_lookup' 'XX' %}";
+                var states = $(stateSelector);
+                states.empty();
+                states.append($('<option>', { value : -1 }).text("State / County / Province / Region"));
```
again, why not use "" instead of -1?

alintreznai replied:
Updated.

sztamas commented on rdrf/registry/groups/registration/base.py:
```diff
@@ -0,0 +1,74 @@
+logger = logging.getLogger(__name__)
+
+
+class BaseRegistration(object):
```
Should be `class BaseRegistration(abc.ABC):` and then `BaseRegistration` can't be initiated because it has @abtractmethods .

alintreznai replied:
Updated.

sztamas commented on rdrf/registry/groups/registration/parent_with_patient/common.py:
```diff
@@ -124,31 +109,54 @@ def process(self):
-            country=request.POST["parent_guardian_country"] if same_address else request.POST["country"]
-        )
-        return address
+    def _create_django_user(self, django_user, registry, is_parent, groups=[], is_clinician=False):
```
This and the correspondent method in the ParentRegistration look suspicious to me.

The name of the method is misleading as it doesn't create the django user. Should be `setup_django_user`?
The mixture of groups and is_parent, is_clinician parameters, when is_parent and is_clinician are done through groups as well seems ugly. Is the groups parameter used at all?

Can we simplify by having a _setup_django_user method instead of this in the base class and maybe do the setting of groups in another method that just takes the group name as an argument?

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/forms/registration_forms.py:
```diff
@@ -0,0 +1,122 @@
+        'phone_number': _('Phone Number')
+    }
+
+    gender_choices = [
```
Could this come from the models so we don't duplicate it here?

alintreznai replied:
Updated.

Commit:
Updates based on code review

alintreznai commented:
Patient registration page problems:
     * header should be Patient registration (instead of User registration) - **Updated**
     * the gender field has list item "bullets" on - **Updated**
     * the gender field error message appears in the wrong place (on the first list item instead above of below the whole list) **Updated**
    * you can submit the form without setting the State - **Updated**
     * the passwords are lost (have to retyped again) if you fail on server-side validation - **Updated**
    * the recaptcha isn't tested server-side? If you wait until the recaptcha expires you can submit the    form without the recaptcha test again, so I'm assuming it isn't tested at all server-side. - **Updated**

Commit:
Verify re-captcha on registration form submit

Commit:
Refactor registration workflow

Commit:
Cosmetical UI changes for the registration pages

Commit:
Fix issue with state selection after form validation

sztamas committed:
State selection after form validation small adjustments.

  - The current value of state can be set from the form (no need to add it
to the template context)
  - Made js function parameter optional

sztamas committed:
Patient registration: more clean-up.

Commit:
Ensure client side form validation for registration, fix issue with gender

Commit:
UI issue fix

Commit:
Remove unused message, fix issue with same address checkbox

Commit:
Handle re-captcha expiration on the client side
